fossas/fossa-cli
# Adds npm Lockfile v3 support (part A)

# Overview

This PR adds initial lockfile v3 support. 
- We only apply this new tactic when we are confident that provided lockfile is v3 (we can parse, and lockFileVersion is explicitly 3)

Out of Scope:
- Nested workspaces (hope to address with part b of pr chain)
- Other resolutions work as before

## Acceptance criteria

- Npm lockfile v3 is supported

## Testing plan

1) Create an example: `package.json`

```
{
    "name": "lockv3",
    "version": "1.0.0",
    "description": "",
    "main": "index.js",
    "scripts": {
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "devDependencies": {
        "mocha": "*"
    },
    "peerDependencies": {
        "chalk": "*"
    },
    "dependencies": {
        "3": "^2.1.0",
        "aws-sdk": "^2.0.0",
        "react": "^18.1.0"
    },
    "author": "",
    "license": "ISC"
}

```

2) Execute: `npm i --lockfile-version=3`
3) Run `fossa-dev analyze -o` and confirm we analyzed it successfully.

## Manual Testing plan
- [x] compare v2 and v3 analysis for a complicated project
- [x] test against symbolic linked package

## To Dos


## Risks

Since lockFile v3 is not used by npm's default configs. This change has very low risk profile.

## References

https://fossa.atlassian.net/browse/ANE-109

## Checklist
